### PR TITLE
Add protection for AddLog_P for ESP32 tasks

### LINKED ISSUE #10049: logging corruption - observed on ESP32 with more that one 'task'
### PROBLEM DESCRIPTION
Whilst developing a new driver (xdrv_47_BLE_ESP32), serial logging corruption was observed.
the specific corruption was something like:
00:00:04 Startscan from on
00:00:04 Startscan from onesec

With ESP32, since multiple 'tasks' are possible (and essential for some operations like BLE if we are not to hold up the whole main process), it is possible to call, for example, Addlog_P from multiple threads.

We need to do some analysis of TasmotaGlobal and the data therein to see what data accesses must be protected by some form of mutex/semaphore when compiling for ESP32, else we will introduce instability and crash and burn badly and randomly.

Immediate danger would be:
Any logging from anything not on the main thread.
Any MQTT activity from anything not on the main thread.

Until we have some form of protection, all logging from secondary threads should be avoided, as should any activity against MQTT or other global data.
One issue is that it may not be obvious when using a 3rd party API like NimBLE if a callback is on the main thread or on a secondary task.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [forum](https://groups.google.com/d/forum/sonoffusers)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): ESP32 dev board
- [ ] Tasmota binary firmware version number used: _____
  - [ ] Pre-compiled
  - [x] Self-compiled
    - [ ] IDE / Compiler used: _____
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```
  STATUS 0 output here:


```
- [ ] Provide the output of the Console log output when you experience your issue; if applicable:
  _(Please use_ `weblog 4` _for more debug information)_
```
  Console output here:


```

### TO REPRODUCE
n/a

### EXPECTED BEHAVIOUR
n/a

### SCREENSHOTS
n/a

### ADDITIONAL CONTEXT
I would assume this only affects ESP32, as ESP8266 does not have multiple threads? (but i'm no expert).

Example of possible Semaphore use:
(what is port_MAX_DELAY anyway? - must check).
```
// this protects our global data, which can be accessed by multiple tasks 
SemaphoreHandle_t  GlobalsMutex;
...
  GlobalsMutex = xSemaphoreCreateMutex();



// example from support.ino
  xSemaphoreTake(GlobalsMutex, portMAX_DELAY);
    memmove(TasmotaGlobal.log_data + strlen(syslog_preamble), TasmotaGlobal.log_data, sizeof(TasmotaGlobal.log_data) - strlen(syslog_preamble));
    TasmotaGlobal.log_data[sizeof(TasmotaGlobal.log_data) -1] = '\0';
    memcpy(TasmotaGlobal.log_data, syslog_preamble, strlen(syslog_preamble));
    PortUdp_write(TasmotaGlobal.log_data, strlen(TasmotaGlobal.log_data));

  xSemaphoreGive(GlobalsMutex); // release mutex

```
The take/give could be a #define, making if compile to nothing when not ESP32?
There are a lot of references to .log_data, .mqtt_data, etc. so this is no small undertaking.

**(Please, remember to close the issue when the problem has been addressed)**

### PR DESCRIPTION
**Pull request by btsimonh**
## Description:
prevents AddLogP from calling other functions if the log call was from a task other than the main task.
Calls PerpLog_P with the format string from the AddLogP call with LOG_LEVEL_ERROR as a guide for diagnosing what called.

**Related issue (if applicable):** fixes #10049

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works on Tasmota core ESP8266 V.2.7.4.7
  - [x] The code change is tested and works on Tasmota core ESP32 V.1.0.4.2
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by arendst** [comment]
I redesigned logging hoping to solve the issue you provide this PR for.

PrepLog_P is gone.

Can you pls verify that the latest developments provide the same function as you hoped to get here?

What changed is that the (global) heap log buffer has been replaced by a (local) stack log buffer solving corruption of log buffer.

**Comment by btsimonh** [comment]
Will take a look.  Just trying to finalise all the BLE stuff at the moment; making sure OTA works, etc.
So I'll drop it over latest dev branch and try the new logging.  a few days?

**Comment by btsimonh** [comment]
ohh... so had a look.
We're going to need mutex protection on TasmotaGlobal.log_buffer and associated control variables.
The implementation using a global buffer with the various log dests pulling from there is nice (more memory efficient, and centralised, compared to my 'SafeAddlog_P' which had a fixed set of constant sized buffer), but if the buffer got screwed once, everything will fall apart really badly....

My next task is to drop my BLE stuff on top of development again, ready for a (hopefully final) PR.  Give me a few days, and I'll propose a global auto-mutex class which we can use in AddLogData and GetLog, which should be the protection we need for ESP32 - and it should be well tested by then too.  I'll drop it as a separate PR if it works....  
I'm not sure about interrupts - I don't even know if the interrupt context supports printf....  we probably should not be logging direct from interrupts except in fatal circumstances, and then only to serial; but I'll defer to the more knowledgeable on that.

**Comment by btsimonh** [comment]
Theo, have a look at https://github.com/btsimonh/Tasmota/tree/loggingmutex two changed files.
The automutex class I added in support.ino can also be used elsewhere, and is only active in esp32.
If you think it's ok, feel free to pull it over.
It's been running here for an hour, no obvious defects on ESP32; but it's based on the code I made for the BLE stuff, so the idea is heavily tested over the last 6 weeks.

**Comment by arendst** [comment]
OK will have a look once I have my SPI project finished which currently breaks stuff. Give me a few days.

**Comment by btsimonh** [comment]
have thrown in a pr.

**Comment by btsimonh** [comment]
@arendst - Theo, ref the logdata being in stack, this adds 700 bytes per thread/task.
1/ Can I move it back to be a static variable, so we only have one of them, protected by the mutex?
2/ You eliminated PrepLog_P - does this mean AddLog_P may be called from interrupt routines now? (i'm scared...)

**Comment by arendst** [comment]
The point of moving logdata to the stack was fixing corruption and no need for a mutex. The advantage was also gaining 700 bytes RAM as most logging messages are way smaller than 700 bytes.

I'll plan and anticipate to make current 700 bytes stack use lower once I have a complete picture of current logging strings; the biggest once, which lead to the 700 size are already handled by their own stack reservations (mqtt logging) so I expect considering saving back to 300 bytes.

The function of PrepLogP was to save logging string before showing it as showing it was the showstopper for both udp and mqtt. The new logging environment does kind of the same so the need for PrepLogP is gone. As it is in use for over a week I haven't seen exceptions on emulation enabled devices caused by this (which was the main culprit).

So in short I do not want to go back to static logbuffer. Certainly not on esp8266.

**Comment by btsimonh** [comment]
sounds good.  I have a small update on the mutex stuff, but the logging will remain unchanged.
Thanks, Simon.  
(p.s. I'm hoping to finish BLE before 'real work' on Monday! :) ).

**Comment by btsimonh** [comment]
closing as really happy with where we got to....
